### ISSUE TITLE
How to detect, when redirect is complete, and all site is loaded ?

### ISSUE THREAD

**Issue by sintanial** [segmentio/nightmare#481]
Hi all.
How to detect, when redirect is complete, and all site is loaded ?
For example, i visit google.com, set some query in input, then press search button, next i choose some site, and click on reference, then browser is loading new site. How to detect when this site is loaded ? 
func "wait" not helped, because i don't know how long site is loading.
Or another example, i click some advertise, which have 5-6 redirects until finish page is loaded, how to detect this ? when finish page is loaded ?
For example zombie.js has eventloop, in which puts events like page loaded, or script evaluated, or css loaded ant etc.... And when i call wait(), then func "wait" waiting until all events in eventloop completed ! 
Is there something like zombie behavior ?

**Comment by geraldyeo** [comment]
I have not tried this, but I believe you can use [.on('did-finish-load', cb)](https://github.com/segmentio/nightmare#onevent-callback)?

**Comment by sintanial** [comment]
yes, i can, but how can i evaluate script after callback is fired ? wait ? wait has bug https://github.com/segmentio/nightmare/issues/251 :(

**Comment by sintanial** [comment]
I created wrapper for wait electron events. I posted this code here, if someone needed:

```jaavascript
// nightmare.js
var _Nightmare = require('nightmare');

var EventEmitter = require("events");
var util = require("util");

var events = [
    'did-finish-load',
    'did-fail-load',
    'did-frame-finish-load',
    'did-start-loading',
    'did-stop-loading',
    'did-get-response-details',
    'did-get-redirect-request',
    'dom-ready',
    'page-favicon-updated',
    'new-window',
    'will-navigate',
    'crashed',
    'plugin-crashed',
    'destroyed'
];

_Nightmare.action('ewait', function (event, cb, done) {
    if (done == undefined) {
        done = cb;
        cb = undefined;
    }

    var result,
        self = this,
        isTimedOut = false,
        tm = null;

    self._proxyEvents.on(event, ()=> {
        if (isTimedOut) return;

        isTimedOut = true;
        clearTimeout(tm);

        if (typeof cb == "function") {
            result = Array.prototype.slice.call(arguments);
            result.unshift(null);
            cb.apply(self, result);
        }

        done();
    });

    tm = setTimeout(()=> {
        if (isTimedOut) return;

        isTimedOut = true;
        clearTimeout(tm);

        var err = new Error('.ewait() timed out after ' + self.optionWaitTimeout + 'msec');

        if (typeof cb == "function") {
            cb.call(self, err);
        }

        done(err);
    }, self.optionWaitTimeout);
});

function ProxyEvents() {
    EventEmitter.call(this);
    this._completed = {};
}

util.inherits(ProxyEvents, EventEmitter);

ProxyEvents.prototype.emit = function () {
    var args = Array.prototype.slice.call(arguments);
    this._completed[args[0]] = args.slice(1);
    EventEmitter.prototype.emit.apply(this, arguments);
};

ProxyEvents.prototype.on = function (event, cb) {
    if (this._completed[event] != undefined) {
        cb.apply(this, this._completed[event]);
        delete this._completed[event];
        return;
    }

    EventEmitter.prototype.on.apply(this, arguments);
};

module.exports = function (options) {
    var nightmare = new _Nightmare(options);

    var _proxyEvents = nightmare._proxyEvents = new ProxyEvents();

    nightmare.on("will-navigate", function () {
        _proxyEvents._completed = {};
    });

    for (var i = 0; i < events.length; i++) {
        (function (i) {
            nightmare.on(events[i], function () {
                var args = Array.prototype.slice.call(arguments);
                _proxyEvents.emit.call(_proxyEvents, events[i], args);
            });
        })(i);
    }

    return nightmare;
};
```

Example for use:
```
var Nightmare = require('./nightmare');
var vo = require('vo');


vo(run)(function (err, result) {
    if (err) throw err;
});

function *run() {
    var nightmare = Nightmare({
        show: true,
        paths: {
            userData: '/dev/null'
        }
    });

    var title = yield nightmare
        .goto("https://github.com/")
        .evaluate(()=> {
            document.querySelector(`.header-actions .btn[href="/login"]`).click();
        })
        .ewait("dom-ready")
        .evaluate(() => {
            document.querySelector("#login_field").value = "{SOME_GITHUB_LOGIN}";
            document.querySelector("#password").value = "{SOME_GITHUB_PASS}";
            document.querySelector(".auth-form-body .btn.btn-primary.btn-block").click();
        })
        .ewait("dom-ready")
        .screenshot("github.png")
        .evaluate(()=> {
            return document.title;
        });
    yield nightmare.end();
    
    console.log(title);
}
```

**Comment by fzzle** [comment]
TLDR, wouldn't this work?
``` coffeescript
Nightmare::ewait = (event) ->
  @queue (done) ->
    @child.once event, ->
      done()
  return @
```

**Comment by navossoc** [comment]
Just a quick fix:
`self.optionWaitTimeout` should be `self.options.waitTimeout`

See: https://github.com/segmentio/nightmare/pull/716/commits/384b752604ed1450a0bda996feba20566dad736f

**Comment by cicerohen** [comment]
@sintanial  thanks for this code example.  Works fine. :D

**Comment by adi518** [comment]
Why do you need this if `evaluate` waits for `did-finish-load`?

**Comment by zeluspudding** [comment]
@sintanial How does one use your wrapper code? Where should it be saved so that I could run your example code?

**Comment by adi518** [comment]
@zeluspudding you should take his code and put it inside a module file, e.g.: `nightmare-ewait.js` . Then, just require it and pass your Nightmare reference so you will have the `ewait` action added to it. However, I think it might be a little outdated/buggy.

**Comment by zeluspudding** [comment]
Thanks @adi518 ! If it is outdated, what's a better solution? I've tried `.on('did-finish-load', cb)?` but it didn't work for some reason
